### PR TITLE
Improve winit and backend selector error handling

### PR DESCRIPTION
**Pull request by tronical**
Propagate the winit error when `DISPLAY` or `WAYLAND_DISPLAY` isn't set and fall back to other available backends.

### CODE REVIEW

File: internal/backends/winit/renderer/sw.rs [hunk 1]
```diff
@@ -69,9 +69,9 @@ impl WinitSoftwareRenderer {
         window_builder: winit::window::WindowBuilder,
     ) -> Result<(Box<dyn WinitCompatibleRenderer>, winit::window::Window), PlatformError> {
         let winit_window = crate::event_loop::with_window_target(|event_loop| {
-            window_builder.build(event_loop.event_loop_target()).map_err(|winit_os_error| {
+            Ok(window_builder.build(event_loop.event_loop_target()).map_err(|winit_os_error| {
                 format!("Error creating native window for software rendering: {}", winit_os_error)
-            })
+            })?)
```
Comment (ogoffart) [inline internal/backends/winit/renderer/sw.rs:74]: I don't understand this change.
It wraps in Ok only to do a `?`?

**Review by ogoffart — APPROVED** [review]

File: internal/backends/selector/lib.rs [hunk 2]
```diff
@@ -46,6 +51,29 @@ cfg_if::cfg_if! {
...
+                create_linuxkms_backend,
+                || Err(PlatformError::NoPlatform),
+            ];
+
+            let mut last_err = PlatformError::NoPlatform;
+
+            for backend_factory in backends {
+                match backend_factory() {
+                    Ok(platform) => return Ok(platform),
+                    Err(err) => last_err = err,
```
Comment (ogoffart) [inline internal/backends/selector/lib.rs:70]: IMHO, this should not return just the last error, but a concatenation of all error. The error should be something like

```
Could not initialize backend.
Error from Qt backend: "Could not connect to display server"
Error from Winit backend: XOpenDisplayFailed
Error from LinuxKMS: could not dlopen shared library "libseat"
```

You get the idea.

**Review by ogoffart — APPROVED** [review]

File: internal/backends/selector/lib.rs (hunk 2 above)
Comment (tronical) [reply internal/backends/selector/lib.rs:70]: Okay, good point. I can do that.

Comment (tronical) [reply internal/backends/selector/lib.rs]: I implemented this now, but it's... sub-optimal:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Other("Could not initialize backend.\nError from Winit backend: Error initializing winit event loop: os error at /home/slint/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.2/src/platform_impl/linux/mod.rs:776: neither WAYLAND_DISPLAY nor DISPLAY is set.\nError from LinuxKMS backend: Timeout while waiting to activate session\nNo backends configured.")', examples/carousel/rust/main.rs:23:23
```

Should I introduce a dedicated PlatformError instead that holds a Vec of other Errors? Or is it perhaps wrong to return an Error here anyway and we should panic?

Comment (ogoffart) [reply internal/backends/selector/lib.rs]: I think it's good for now.

We may want to improve the Debug for our error maybe.
But that can be done another time.

**Review by ogoffart — APPROVED** [review]

**Pull request merged**